feat(btcstaking): committing/timestamping public randomness via BIP-32

SebastianElvis opened the pull request:
Fixes #591 

This PR provides the full implementation on deriving/timestamping EOTS public randomness via BIP-32. It consists of a series of small PRs (see below commit history).

NOTE: this should be merged via a merge commit

gitferry commented on x/checkpointing/keeper/keeper.go:
```diff
@@ -395,3 +399,29 @@ func (k Keeper) GetTotalVotingPower(ctx context.Context, epochNumber uint64) int
+// GetLastFinalizedEpoch gets the last finalised epoch
+func (k Keeper) GetLastFinalizedEpoch(ctx context.Context) (uint64, error) {
+	store := k.storeService.OpenKVStore(ctx)
+	has, err := store.Has(types.LastFinalizedEpochKey)
+	if err != nil {
+		panic(err)
+	}
+	if !has {
+		return 0, types.ErrFinalizedEpochNotFound
+	}
+	epochNumberBytes, err := store.Get(types.LastFinalizedEpochKey)
+	if err != nil {
+		panic(err)
+	}
```
Seems that `store.Get` is sufficient for our purpose. We can do nil check over `epochNumberBytes` to return `types.ErrFinalizedEpochNotFound`. Also, I think maybe we can panic on `types.ErrFinalizedEpochNotFound` if we decide to set epoch `0` finalized

gitferry commented on x/btcstaking/client/cli/tx.go:
```diff
@@ -52,8 +52,8 @@ func GetTxCmd() *cobra.Command {
 
 func NewCreateFinalityProvicerCmd() *cobra.Command {
```
```suggestion
func NewCreateFinalityProviderCmd() *cobra.Command {
```

gitferry commented on x/finality/keeper/msg_server.go:
```diff
@@ -77,6 +77,15 @@ func (ms msgServer) AddFinalitySig(goCtx context.Context, req *types.MsgAddFinal
 	}
 
+	// ensure the finality provider's registered epoch is already finalised by BTC timestamping
+	finalizedEpoch, err := ms.BTCStakingKeeper.GetLastFinalizedEpoch(ctx)
```
what if the finality sig is sent but epoch `1` has not been finalized? Will this lead to panic?

gitferry reviewed (approved):
Very neat set of prs! Only question is about finalizing epoch `0`

SebastianElvis commented:
> Very neat set of prs! Only questions are around finalizing epoch `0`

Great point on finalising epoch 0! I have addressed the comment by finalising epoch 0 at genesis at https://github.com/babylonchain/babylon/pull/606. Please feel free to take a look

SebastianElvis commented on x/checkpointing/genesis.go:
```diff
@@ -10,6 +11,8 @@ import (
 func InitGenesis(ctx context.Context, k keeper.Keeper, genState types.GenesisState) {
 	k.SetGenBlsKeys(ctx, genState.GenesisKeys)
+	// set epoch 0 to be finalised at genesis
+	k.SetLastFinalizedEpoch(ctx, 0)
```
Good point. Looks like the genesis import/export is not supported in the checkpointing module. This not only includes the last finalsied epoch, but also other storage, e.g., checkpoints. Left a TODO at https://github.com/babylonchain/babylon/pull/609 for now and need to address it in a separate PR
